File: chatbot/flow_manager.py
#!/usr/bin/env python3
"""
TextIt Flow Manager
Gestiona flujos de chatbot usando la API oficial de RapidPro
"""

import sys
import json
from temba_client.v2 import TembaClient

# Configuración
API_TOKEN = 'bfe6fe930e078ddfeac32f8ebccbf2428d0f43e0'
API_URL = 'https://textit.com'

# Cliente de la API
client = TembaClient(API_URL, API_TOKEN)

def list_flows():
    """Lista todos los flujos disponibles"""
    print("📋 Listando flujos...\n")
    
    flows = client.get_flows().all()
    
    for i, flow in enumerate(flows, 1):
        print(f"{i}. {flow.name}")
        print(f"   UUID: {flow.uuid}")
        print(f"   Tipo: {flow.type}")
        print(f"   Archivado: {'Sí' if flow.archived else 'No'}")
        print(f"   Ejecuciones completadas: {flow.runs.completed}")
        print()

def download_flow(uuid):
    """Descarga la definición completa de un flujo"""
    if not uuid:
        print("❌ Error: Debes proporcionar el UUID del flujo")
        print("Uso: python3 flow_manager.py download <UUID>")
        return
    
    print(f"📥 Descargando flujo {uuid}...")
    
    # Obtener definición del flujo
    definitions = client.get_definitions(flows=[uuid])
    
    if definitions.flows:
        flow_def = definitions.flows[0]
        filename = f"flows/{flow_def['name'].replace(' ', '_')}_{uuid[:8]}.json"
        
        # Crear directorio si no existe
        import os
        os.makedirs('flows', exist_ok=True)
        
        # Guardar JSON
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(flow_def, f, indent=2, ensure_ascii=False)
        
        print(f"✅ Flujo guardado en: {filename}")
        print(f"   Nombre: {flow_def['name']}")
        print(f"   Nodos: {len(flow_def.get('nodes', []))}")
    else:
        print("❌ No se encontró el flujo")

def upload_flow(filepath):
    """Sube/actualiza un flujo desde un archivo JSON"""
    if not filepath:
        print("❌ Error: Debes proporcionar la ruta al archivo JSON")
        print("Uso: python3 flow_manager.py upload <archivo.json>")
        return
    
    print(f"📤 Subiendo flujo desde {filepath}...")
    
    try:
        import requests
        
        with open(filepath, 'r', encoding='utf-8') as f:
            flow_def = json.load(f)
        
        # Importar el flujo usando flow_definitions endpoint
        url = f"{API_URL}/api/v2/flow_definitions.json"
        headers = {
            'Authorization': f'Token {API_TOKEN}',
            'Content-Type': 'application/json'
        }
        
        # El payload debe ser un objeto con flows array
        payload = {
            'flows': [flow_def]
        }
        
        response = requests.post(url, json=payload, headers=headers)
        
        if response.status_code in [200, 201]:
            result = response.json()
            print(f"✅ Flujo importado exitosamente")
            print(f"   Resultado: {result}")
        else:
            print(f"❌ Error al subir el flujo: HTTP {response.status_code}")
            print(f"   Respuesta: {response.text}")
        
    except Exception as e:
        print(f"❌ Error al subir el flujo: {e}")

def search_text(text):
    """Busca texto en flujos locales"""
    if not text:
        print("❌ Error: Debes proporcionar el texto a buscar")
        print("Uso: python3 flow_manager.py search <texto>")
        return
    
    print(f"🔍 Buscando '{text}' en flujos locales...\n")
    
    import os
    import glob
    
    if not os.path.exists('flows'):
        print("⚠️  No hay flujos descargados. Usa: python3 flow_manager.py download <UUID>")
        return
    
    found = 0
    for filepath in glob.glob('flows/*.json'):
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
            
        if text.lower() in content.lower():
            found += 1
            print(f"📄 {os.path.basename(filepath)}")
            
            # Mostrar contexto
            flow = json.loads(content)
            for i, node in enumerate(flow.get('nodes', [])):
                for action in node.get('actions', []):
                    if 'text' in action and text.lower() in action['text'].lower():
                        preview = action['text'][:80] + "..." if len(action['text']) > 80 else action['text']
                        print(f"   Nodo {i}: \"{preview}\"")
            print()
    
    if found == 0:
        print("❌ No se encontraron coincidencias")
    else:
        print(f"✅ Encontrado en {found} archivo(s)")

def show_help():
    """Muestra la ayuda"""
    print("""
📚 TextIt Flow Manager - Comandos disponibles:

  list                    Lista todos los flujos disponibles
  download <UUID>         Descarga un flujo específico
  upload <archivo.json>   Sube/actualiza un flujo
  search <texto>          Busca texto en flujos locales
  help                    Muestra esta ayuda

Ejemplos:
  python3 flow_manager.py list
  python3 flow_manager.py download e3837ec8-d0c7-4e9b-a14a-6fdcc24e1e13
  python3 flow_manager.py upload flows/mi_flujo.json
  python3 flow_manager.py search "Registro Exitoso"
    """)

# Main
if __name__ == '__main__':
    commands = {
        'list': list_flows,
        'download': lambda: download_flow(sys.argv[2] if len(sys.argv) > 2 else None),
        'upload': lambda: upload_flow(sys.argv[2] if len(sys.argv) > 2 else None),
        'search': lambda: search_text(sys.argv[2] if len(sys.argv) > 2 else None),
        'help': show_help
    }
    
    command = sys.argv[1] if len(sys.argv) > 1 else 'help'
    
    if command in commands:
        commands[command]()
    else:
        print(f"❌ Comando desconocido: {command}")
        show_help()

File: chatbot/flow-manager.js
#!/usr/bin/env node

/**
 * TextIt Flow Manager
 * Gestiona flujos de chatbot desde la línea de comandos
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

const API_TOKEN = 'bfe6fe930e078ddfeac32f8ebccbf2428d0f43e0';
const API_BASE = 'textit.com';

// Helper para hacer requests a la API
function apiRequest(endpoint, method = 'GET', data = null) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: API_BASE,
      path: `/api/v2${endpoint}`,
      method: method,
      headers: {
        'Authorization': `Token ${API_TOKEN}`,
        'Content-Type': 'application/json'
      }
    };

    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', chunk => body += chunk);
      res.on('end', () => {
        try {
          const json = JSON.parse(body);
          resolve(json);
        } catch (e) {
          resolve(body);
        }
      });
    });

    req.on('error', reject);
    
    if (data) {
      req.write(JSON.stringify(data));
    }
    
    req.end();
  });
}

// Comandos disponibles
const commands = {
  // Listar todos los flujos
  async list() {
    console.log('📋 Listando flujos...\n');
    const response = await apiRequest('/flows.json');
    
    if (!response.results || response.results.length === 0) {
      console.log('⚠️  No se encontraron flujos o hubo un error en la API');
      console.log('Response:', response);
      return;
    }
    
    response.results.forEach((flow, i) => {
      console.log(`${i + 1}. ${flow.name}`);
      console.log(`   UUID: ${flow.uuid}`);
      console.log(`   Tipo: ${flow.type}`);
      console.log(`   Archivado: ${flow.archived ? 'Sí' : 'No'}`);
      console.log(`   Ejecuciones completadas: ${flow.runs.completed}`);
      console.log('');
    });
  },

  // Descargar definición de un flujo
  async download(uuid) {
    if (!uuid) {
      console.error('❌ Error: Debes proporcionar el UUID del flujo');
      console.log('Uso: node flow-manager.js download <UUID>');
      return;
    }

    console.log(`📥 Descargando flujo ${uuid}...`);
    const response = await apiRequest(`/flow_definitions.json?uuid=${uuid}`);
    
    if (response.results && response.results.length > 0) {
      const flow = response.results[0];
      const filename = `${flow.name.replace(/[^a-z0-9]/gi, '_')}_${uuid.substring(0, 8)}.json`;
      const filepath = path.join(__dirname, 'flows', filename);
      
      // Crear directorio flows si no existe
      if (!fs.existsSync(path.join(__dirname, 'flows'))) {
        fs.mkdirSync(path.join(__dirname, 'flows'));
      }
      
      fs.writeFileSync(filepath, JSON.stringify(flow, null, 2));
      console.log(`✅ Flujo guardado en: flows/${filename}`);
    } else {
      console.error('❌ No se encontró el flujo');
    }
  },

  // Subir/actualizar un flujo
  async upload(filepath) {
    if (!filepath) {
      console.error('❌ Error: Debes proporcionar la ruta al archivo JSON');
      console.log('Uso: node flow-manager.js upload <archivo.json>');
      return;
    }

    console.log(`📤 Subiendo flujo desde ${filepath}...`);
    
    const flowData = JSON.parse(fs.readFileSync(filepath, 'utf8'));
    const response = await apiRequest('/flow_definitions.json', 'POST', flowData);
    
    if (response.uuid) {
      console.log(`✅ Flujo actualizado exitosamente`);
      console.log(`   UUID: ${response.uuid}`);
      console.log(`   Nombre: ${response.name}`);
    } else {
      console.error('❌ Error al subir el flujo:', response);
    }
  },

  // Buscar texto en flujos locales
  async search(text) {
    if (!text) {
      console.error('❌ Error: Debes proporcionar el texto a buscar');
      console.log('Uso: node flow-manager.js search <texto>');
      return;
    }

    console.log(`🔍 Buscando "${text}" en flujos locales...\n`);
    
    const flowsDir = path.join(__dirname, 'flows');
    if (!fs.existsSync(flowsDir)) {
      console.log('⚠️  No hay flujos descargados. Usa: node flow-manager.js download <UUID>');
      return;
    }

    const files = fs.readdirSync(flowsDir).filter(f => f.endsWith('.json'));
    let found = 0;

    files.forEach(file => {
      const content = fs.readFileSync(path.join(flowsDir, file), 'utf8');
      if (content.toLowerCase().includes(text.toLowerCase())) {
        found++;
        console.log(`📄 ${file}`);
        
        // Mostrar contexto
        const flow = JSON.parse(content);
        flow.nodes?.forEach((node, i) => {
          node.actions?.forEach(action => {
            if (action.text?.toLowerCase().includes(text.toLowerCase())) {
              console.log(`   Nodo ${i}: "${action.text.substring(0, 80)}..."`);
            }
          });
        });
        console.log('');
      }
    });

    if (found === 0) {
      console.log('❌ No se encontraron coincidencias');
    } else {
      console.log(`✅ Encontrado en ${found} archivo(s)`);
    }
  },

  // Ayuda
  help() {
    console.log(`
📚 TextIt Flow Manager - Comandos disponibles:

  list                    Lista todos los flujos disponibles
  download <UUID>         Descarga un flujo específico
  upload <archivo.json>   Sube/actualiza un flujo
  search <texto>          Busca texto en flujos locales
  help                    Muestra esta ayuda

Ejemplos:
  node flow-manager.js list
  node flow-manager.js download e3837ec8-d0c7-4e9b-a14a-6fdcc24e1e13
  node flow-manager.js upload flows/mi_flujo.json
  node flow-manager.js search "Registro Exitoso"
    `);
  }
};

// Ejecutar comando
const [,, command, ...args] = process.argv;

if (!command || !commands[command]) {
  commands.help();
} else {
  commands[command](...args).catch(console.error);
}
